# Reference:
#   - https://github.com/actions/checkout
#   - https://github.com/actions/upload-artifact
#   - https://github.com/prefix-dev/setup-pixi
#   - https://github.com/pyvista/setup-headless-display-action

name: ci-tests-lock

on:
  workflow_dispatch:
  schedule:
    # every day @ 00h03 except sunday (see ci-locks.yml)
    - cron: "3 0 * * 1-6"

concurrency:
  group: ${{ github.workflow }}-${{ github.ref }}
  cancel-in-progress: true

permissions:
  issues: write

defaults:
  run:
    shell: bash -l {0}

jobs:
  test-lock:
    name: "test lock (${{ matrix.version }})"
    runs-on: "${{ matrix.os }}"

    env:
      ALLOW_PLOTTING: true
      PY_COLORS: "1"

    strategy:
      fail-fast: false
      matrix:
        os: ["ubuntu-22.04"]
        version: ["py313"]

    steps:
      - uses: actions/checkout@v5
        with:
          fetch-depth: 0
          persist-credentials: false

      - name: "environment configure"
        env:
          # Maximum cache period (in weeks) before forcing a cache refresh.
          CACHE_WEEKS: 2
        run: |
          echo "CACHE_PERIOD=$(date +%Y).$(expr $(date +%U) / ${CACHE_WEEKS})" >> ${GITHUB_ENV}
          echo "NAME=geovista-${{ matrix.version }}" >> ${GITHUB_ENV}

      - uses: pyvista/setup-headless-display-action@7d84ae825e6d9297a8e99bdbbae20d1b919a0b19

      - name: "setup pixi"
        uses: prefix-dev/setup-pixi@fef5c9568ca6c4ff7707bf840ab0692ba3f08293
        with:
          cache-write: false
          run-install: false

      - name: "refresh pixi manifest"
        run: |
          pixi update --json | pixi exec pixi-diff-to-markdown > diff.md

      - name: "upload artifacts"
        uses: actions/upload-artifact@v4
        with:
          name: lock-artifacts-${{ github.job }}-${{ strategy.job-index }}
          path: |
            ${{ github.workspace }}/diff.md
            ${{ github.workspace }}/pixi.lock

      - name: "cartopy cache"
        uses: ./.github/workflows/composite/cartopy-cache
        with:
          cache_period: ${{ env.CACHE_PERIOD }}

      - name: "geovista cache"
        uses: ./.github/workflows/composite/geovista-cache
        with:
          cache_period: ${{ env.CACHE_PERIOD }}
          pixi_environment: ${{ env.NAME }}

      - name: "test lock (${{ matrix.version }})"
        env:
          NAME: ${{ env.NAME }}
        run: |
          xvfb-run -a pixi run --frozen --environment ${NAME} pytest --generated_image_dir test_images

      - name: "report failure"
        if: failure()
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          NAME: ${{ env.NAME }}
        run: |
          TITLE="chore: pixi.lock test failure"
          BODY=$(cat <<EOL
          The \`${{ matrix.version }}\` tests for the latest resolved \`pixi.lock\` have failed for the \`${NAME}\` environment.

          Please investigate GHA run-id [${{ github.run_id }}](https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }})
          EOL
          )
          gh issue create --title "${TITLE}" --body "${BODY}" --label "bot" --label "new: issue" --repo "${{ github.repository }}"
